File: backend/mallapi/src/test/java/org/nareun130/mallapi/repository/ProductRepositoryTests.java
package org.nareun130.mallapi.repository;

import java.util.Arrays;
import java.util.Optional;
import java.util.UUID;

import org.junit.jupiter.api.Test;
import org.nareun130.mallapi.domain.Product;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.test.annotation.Commit;
import org.springframework.transaction.annotation.Transactional;

import lombok.extern.log4j.Log4j2;

@SpringBootTest
@Log4j2
public class ProductRepositoryTests {
    
    @Autowired
    ProductRepository productRepository;

    @Test
    public void testInsert() {

        for(int i = 0;  i < 10; i++) {

            Product product = Product.builder()    
                .pname("상품" + i)
                .price(100*i) 
                .pdesc("상품설명 " + i)
                .build();

            //2개의 이미지 파일을 추가
            product.addImageString(UUID.randomUUID().toString()+"_"+"IMAGE1.jpg");
            product.addImageString(UUID.randomUUID().toString()+"_"+"IMAGE2.jpg");

            productRepository.save(product);

            log.info("-------------------------------");
        }

        
    }

    @Transactional //* db를 2번 타야해서 Transactional 처리
    @Test
    public void testRead() {

        Long pno = 1L;

        Optional<Product> result = productRepository.findById(pno);

        Product product = result.orElseThrow();

        log.info(product);
        log.info(product.getImageList());

    }

    @Test // ~> DB를 한번만 타도록 
    public void testRead2() {

        Long pno = 1L;

        Optional<Product> result = productRepository.selectOne(pno);

        Product product = result.orElseThrow();

        log.info(product);
        log.info(product.getImageList());

    }

    @Commit // * test 후 커밋하도록
    @Transactional
    @Test
    public void testDelete() {
        Long pno = 2L;

        productRepository.updateToDelete(pno, true);
    }

    @Test
    public void testUpdate() {

        Long pno = 10L;

        Product product = productRepository.selectOne(pno).get();

        product.changeName("10번 상품");
        product.changeDesc("10번 상품 설명입니다.");
        product.changePrice(5000);

        product.clearList();

        product.addImageString(UUID.randomUUID().toString() + "_" + "NEWIMAGE1.jpg");
        product.addImageString(UUID.randomUUID().toString() + "_" + "NEWIMAGE2.jpg");
        product.addImageString(UUID.randomUUID().toString() + "_" + "NEWIMAGE3.jpg");
        
        productRepository.save(product);
    }

    @Test
    public void testList() {

        Pageable pageable = PageRequest.of(0, 10 , Sort.by("pno").descending());

        Page<Object[]> result = productRepository.selectList(pageable);

        result.getContent().forEach(arr -> log.info(Arrays.toString(arr)));
    }
}

File: backend/mallapi-kotlin/src/test/kotlin/com/nareun/mallapikotlin/repository/ProductRepositoryTests.kt
package com.nareun.mallapikotlin.repository

import com.nareun.mallapikotlin.domain.Product
import lombok.extern.log4j.Log4j2
import org.junit.jupiter.api.Test
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.annotation.Commit
import org.springframework.transaction.annotation.Transactional
import java.util.*


@SpringBootTest
@Log4j2
class ProductRepositoryTests {

    @Autowired
    lateinit var productRepository: ProductRepository

    val log = LoggerFactory.getLogger(ProductRepositoryTests::class.java)
    @Test
    fun testInsert() {
        repeat(10) { i ->
            val product = Product(
                pname = "상품$i",
                price = 100 * i,
                pdesc = "상품설명 $i"
            ).apply {
                //둘 다 사용 가능
                addImageString(UUID.randomUUID().toString()+"_"+"IMAGE1.jpg")
                addImageString("${UUID.randomUUID()}_IMAGE2.jpg")
            }
            productRepository.save(product)
            log.info("------------------")

        }
    }

    @Transactional //상품 정보를 가져오고 상품 이미지 리스트를 가져오므로 2번 쿼리를 타기 때문 -> Transactional
    @Test
    fun testRead() {
        val pno = 1L

        val product = productRepository.findById(pno).orElseThrow()

        log.info(product.toString())
        log.info(product.imageList.toString())
    }

    @Test
    fun testRead2(){
        val pno : Long = 1L
        val result : Optional<Product> = productRepository.selectOne(pno);

        val product : Product = result.orElseThrow()

        log.info(product.toString())
        log.info(product.imageList.toString())
    }

    @Commit// @Transactional이 기본적으로 롤백 처리를 하도록 설계되어있음.
    // -> 테스트 코드 실행 성공 후 업데이트 된 결과를 DB에서 확인하기 위해 @Commit사용
    @Transactional
    @Test
    fun testDelete(){
        val pno:Long = 2L;

        productRepository.updateToDelete(pno,true)
    }

    @Test
    fun testUpdate(){
        val pno = 10L

        val product :Product = productRepository.selectOne(pno).get()

        product.changeName("10번 상품")
        product.changeDesc("10번 상품 설명입니다.")
        product.changePrice(5000)

        //첨부파일 수정
        product.clearList()

        product.addImageString("${UUID.randomUUID()}_NEWIMAGE1.jpg")
        product.addImageString("${UUID.randomUUID()}_NEWIMAGE2.jpg")
        product.addImageString("${UUID.randomUUID()}_NEWIMAGE3.jpg")

        productRepository.save(product)
    }
}